enhancements/oc/mirroring-configuration-manifests.md

```diff
@@ -0,0 +1,254 @@
+# Mirroring Configuration Manifests
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+For disconnected clusters, `oc adm release mirror ...` already helps users copy the release image and its dependencies into their local mirror.
+This enhancment extends that command to also create and apply configuration manifests to their local cluster, to bring in information that cannot be represented by container images.
+For example, `oc` might produce ConfigMap manifests containing the release image signature which the cluster-version operator could use to [verify the mirrored release][cvo-config-map-signatures].
+
+## Motivation
+
+There is a lot of information in the wider internet that may be useful for restricted-network clusters.
+Some of that information can be represented by container images, such as the release image and the operator and other images referenced by the release image.
+Some of that information cannot be represented by container images, including container signatures.
+By extending `oc`'s release-mirror support to create and apply Kubernetes configuration manifests, we provide a channel that can be used to gather arbitrary information in one place that the user can audit, transport into their restricted network, and apply to their local cluster.
+As the need for additional information arises, additional configuration manifests may be created, possibly using new [custom resources][custom-resources] known to the local cluster, without users needing to adjust their mirroring workflows.
+
+### Goals
+
+Users should be able to run a single `oc adm release mirror ...` command to gather all the external information needed to deploy that release inside their cluster, which may have restricted network access.
+
+### Non-Goals
+
+This proposal does not specify the types of configuration manifests that are created.
+Users who intend to audit the gathered configuration manifests should extract the gathered types from the configuration manifests themselves.
+
+## Proposal
+
+### Applying directly to the target cluster
+
+As of 4.2 and 4.3:
+
+```console
+$ oc adm release mirror \
+>   --from=quay.io/openshift-release-dev/ocp-release:4.3.0-x86_64 \
+>   --to=registry.example.com/your/repository \
+>   --to-release-image=registry.example.com/your/repository:your-tag
+```
+
+will mirror the 4.3.0 release image and other images referenced by the release image into the mirror repository at registry.example.com.
+
+This enhancement adds an `--apply-config` option.
+When set, `oc` would also apply associated configuration manifests to the cluster that `oc` connects to, with the manifest application beginning after the image mirror had completed.
```

Comment:
Where do these manifests come from? How would Red Hat make them available, how would oc know where to find them, and how are they life-cycled?

Reply:
> How would Red Hat make them available, how would oc know where to find them...

`oc` will create them, see openshift/oc#343.

> ... and how are they life-cycled?

As discussed in the enhancement, we need to restrict ourselves to situations where the cluster can function smoothly with or without a given manifest.  So cluster admins could just remove manifests that aren't needed anymore.  If `oc` changes its mind about a manifest it previously generated, we might end up with a mirror operator (or piggy back on the CVO) to remove existing mirror manifests that are no longer useful.  For signatures specifically, they are small enough and few enough that we'll probably just feed them into clusters with no garbage collection in the near- and mid-term.  But these are good points; I'll update the enhancement to mention them.

Reply:
I think I'm just missing some basic background info. What source information does oc use to decide what goes into the manifests? Is it just going to grab signatures directly from here? https://mirror.openshift.com/pub/openshift-v4/signatures/openshift/release/

Reply:
The release image itself, the one being mirrored, contains the signature manifests so it gets pulled from there.

Reply:
I'm sure I'm just missing something basic; if the signature is in the release itself, and the release is already being mirrored, why can't the end cluster pull out the signature data at update time? Why do we need to pre-deliver the signatures to the cluster?

Reply:
Sorry, I stopped too soon. oc is reusing the same same code CVO uses to do the verify and yes, it does go out and download the signature data. The end cluster may be air-gaped from the internet or otherwise unable to access the signatures directly.

Reply:
Ok, so is this proposal suggesting that `oc` will retrieve signatures from mirror.redhat.com and use them to create a manifest that includes a ConfigMap with signature data?

Reply:
> Ok, so is this proposal suggesting...

Yup, that's the plan (and already stubbed into openshift/oc#343).

Reply:
> ... I'll update the enhancement to mention them.

I've pushed 7dcc2b9 -> 884166a, adding a paragraph about the lack of garbage collection and potential future mitigations if we start to get worried about bloat.